flybywiresim/aircraft
# feat(a380x/mfd): whitelist keyboard inputs to keys present on the KCCU

Replaces #9119

## Summary of Changes

This PR does two things:
1. When using your (physical) computer keyboard to enter data into inputs in the MFD, we previously allowed any keys to be pressed, even keys that aren't on the KCCU. This resulted in visual oddities such as unsupported characters being rendered as whitespace and lead to data entering the FMS that never could.
2. With many keys now being rejected, we can helpfully remap keys to more appropriate ones. For example, as outlined in #9119, there's many countries in the world where comma is the decimal separator. As comma is now no longer a valid entry, we can assume that the user wanted to enter a decimal separator and just write a dot instead. This makes the entering data via the (physical) numpad much more convenient when using a keyboard layout such as e.g. German.

## Screenshots (if necessary)


Pressing comma without this PR:
![378898861-37328536-55c5-4ee6-bc96-3bb0c48489a2](https://github.com/user-attachments/assets/3b4ea53e-7c7c-4e3e-977e-d34b29b1bf02)
Pressing comma with this PR:
![378898616-70879a73-d0a5-417a-82ef-2cb787b8a4b7](https://github.com/user-attachments/assets/1b479ec8-0b18-4976-b108-36f8a5414282)

## References





## Additional context



Discord username (if different from GitHub):

## Testing instructions

1. Focus any input on the MFD (e.g. FMS 1/2 > POSITION > NAVAIDS > FREQ
2. Enter all the keys present on the KCCU via your keyboard. It should result in the same experience as if you had used the KCCU. **Note:** The whitespace key (`SP` on the KCCU) does not seem to work. This is unrelated to this PR.
3. Enter some keys that are *not* present on the KCCU (with the exception of comma). Nothing should happen.
4. Set your Windows keyboard layout to one where the decimal separator is a comma (e.g. German).
5. Enter a decimal number entirely via the numpad on your keyboard. You should notice the decimal key being consistenly entered as a dot.


## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, find and click on the **PR Build** tab
1. Click on either **flybywire-aircraft-a320-neo** or **flybywire-aircraft-a380-842** download link at the bottom of the page
